src/main/java/com/ita/if103java/ims/dao/impl/TransactionDaoImpl.java

```diff
@@ -0,0 +1,135 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.dao.DataAccessException;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
+import org.springframework.jdbc.support.GeneratedKeyHolder;
+import org.springframework.jdbc.support.KeyHolder;
+import org.springframework.stereotype.Repository;
+
+import javax.sql.DataSource;
+import java.math.BigInteger;
+import java.sql.*;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+@Repository
+public class TransactionDaoImpl implements TransactionDao {
+    private static Logger LOGGER = LoggerFactory.getLogger(TransactionDaoImpl.class);
+    private TransactionRowMapper mapper;
+    private JdbcTemplate jdbcTemplate;
+    private NamedParameterJdbcTemplate namedParameterJdbcTemplate;
+
+    @Autowired
+    public TransactionDaoImpl(DataSource dataSource, TransactionRowMapper mapper) {
+        this.jdbcTemplate = new JdbcTemplate(dataSource);
+        this.namedParameterJdbcTemplate = new NamedParameterJdbcTemplate(dataSource);
+        this.mapper = mapper;
+    }
+
+    @Override
+    public Transaction create(Transaction transaction) throws DataAccessException {
+        return withErrorLogging(() -> {
+            final KeyHolder keyHolder = new GeneratedKeyHolder();
+            jdbcTemplate.update(connection -> getPreparedStatement(connection, transaction), keyHolder);
+            transaction.setId((BigInteger) keyHolder.getKey());
+            return transaction;
+        }, "Transaction creation failure => " + transaction);
+    }
+
+    @Override
+    public Transaction findById(BigInteger id) throws DataAccessException {
+        return withErrorLogging(() -> jdbcTemplate.queryForObject(Queries.SQL_SELECT_TRANSACTION_BY_ID, mapper, id),
+            "Transaction error on 'select by id' => id=" + id);
+    }
+
+    @Override
+    public List<Transaction> findAll(Integer offset, Integer limit) throws DataAccessException {
+        return withErrorLogging(() -> jdbcTemplate.query(Queries.SQL_SELECT_ALL_TRANSACTIONS, mapper, limit, offset),
+            "Transaction error on 'select *'");
+    }
+
+    @Override
+    public List<Transaction> findAll(Map<String, Object> params,
+                                     Integer offset, Integer limit,
+                                     String orderBy) throws DataAccessException, IllegalStateException {
+        return withErrorLogging(() -> {
+                final String where = Stream
+                    .of("account_id", "associate_id", "item_id", "quantity", "moved_from", "moved_to", "type")
+                    .filter(params::containsKey)
+                    .map(x -> String.format("%s %s :%s", x, params.get(x) == null ? "is" : "=", x))
```

Comment:
not sure what is 'is', could you please explain?
and the last argument, I think, should be `params.get(x)` isn't it?